### ISSUE TITLE
romio: Inconsistent Use of Communicator in ADIOI_GEN_OpenColl for ADIOI_xxx_Close

### ISSUE THREAD

**Issue by range3** [pmodels/mpich#6868]
https://github.com/pmodels/mpich/blob/5d478d9a76d970ad198ea3f2126c5b13e36f57e1/src/mpi/romio/adio/common/ad_opencoll.c#L80-L87

In the implementation of ADIOI_GEN_OpenColl, if the ADIO_CREATE flag is set, ranklist[0] is representing to create the file.
During this, the file is opened for creation with MPI_COMM_SELF, but for closing, it is called with the communicator that was used when ADIOI_GEN_OpenColl was called.

Apologies for Uncertainty, I do not know the detailed specification of ADIO's implementation, so I am not sure,
but since MPI_File_close is collective, I think it is correct to pass the same communicator, MPI_COMM_SELF, as used during open.

**Comment by wkliao** [comment]
MPI_File_close  is indeed collective, but ADIOI_xxx_Close may not. ADIOI_xxx_Close is an internal implementation in ROMIO for individual processes to close a file.

I just now did a quick check and found that all ADIO drivers are not doing communication in their ADIOI_xxx_Close, except for DAOS, in which `ADIOI_DAOS_Close` does call MPI_Barrier which can cause a program to hang. (I don't know why DAOS driver requires a barrier there, and am wondering why DAOS users have not complained yet.)
https://github.com/pmodels/mpich/blob/1f055eeb0443b9349c663e9679b5a40dfd7cc2a8/src/mpi/romio/adio/ad_daos/ad_daos_close.c#L15

@range3, It will be great if you can create a pull request to fix this problem.

**Comment by hzhou** [comment]
> I just now did a quick check and found that all ADIO drivers are not doing communication in their ADIOI_xxx_Close, except for DAOS, in which ADIOI_DAOS_Close does call MPI_Barrier which can cause a program to hang. (I don't know why DAOS driver requires a barrier there, and am wondering why DAOS users have not complained yet.)

Pinging @mchaarawi

**Comment by mchaarawi** [comment]
> > I just now did a quick check and found that all ADIO drivers are not doing communication in their ADIOI_xxx_Close, except for DAOS, in which ADIOI_DAOS_Close does call MPI_Barrier which can cause a program to hang. (I don't know why DAOS driver requires a barrier there, and am wondering why DAOS users have not complained yet.)
> 
> Pinging @mchaarawi

right, The barrier is not needed there and can be removed. I do not get though why it would cause the program to hang. All ranks in comm1 that call MPI_File_open(comm1) do call this function on file close in my experience. otherwise none of our MPIO testing would complete and as you said would hang..

**Comment by wkliao** [comment]
The code fragment pointed by @range3 runs when creating a new file.
In MPI_File_open, the file creation is done by `rank == fd->hints->ranklist[0]` only.
If the creation is successful, then that rank will close the file, and later the file will be
opened by all without creation flag.

File close at line 87 is called only by ranklist[0], but `comm1` is used to call MPI_Barrier
in ADIOI_DAOS_Close, which should be MPI_COMM_SELF.

Can you try a test program that creates a new file to see if it hangs?

**Comment by range3** [comment]
I'm going to create a PR today.
Thank you for reply.

**Comment by range3** [comment]
https://github.com/pmodels/mpich/blob/846075e6abe5039f2e03cf41df8758a0b6b5c1bb/src/mpi/romio/adio/ad_daos/ad_daos.c#L13
DAOS uses ADIOI_DAOS_OpenColl and not ADIOI_GEN_OpenColl, so I think it is not a problem.
